packages/backend/src/cache/middleware.ts

```diff
@@ -0,0 +1,118 @@
+import isEqual from 'lodash/isEqual'
+import { Logger } from 'pino'
+import { CacheDataStore } from './data-stores'
+
+interface CachedRequest {
+  requestResult: unknown
+  requestParams: Record<string, unknown>
+  operationName: string
+}
+
+type Request = () => Promise<unknown>
+
+interface CacheMiddlewareArgs {
+  deps: { logger: Logger; dataStore: CacheDataStore }
+  idempotencyKey?: string
+  request: Request
+  requestParams: Record<string, unknown>
+  operationName: string
+  handleParamMismatch: () => never
+}
+
+export async function cacheMiddleware(
+  args: CacheMiddlewareArgs
+): ReturnType<Request> {
+  const {
+    deps: { dataStore },
+    idempotencyKey,
+    request
+  } = args
+  if (!idempotencyKey) {
+    return request()
+  }
+
+  const cachedRequestAsString = await dataStore.get(idempotencyKey)
+
+  return cachedRequestAsString
+    ? handleCacheHit({ ...args, idempotencyKey }, cachedRequestAsString)
+    : handleCacheMiss({ ...args, idempotencyKey })
+}
+
+async function handleCacheHit(
+  args: Required<CacheMiddlewareArgs>,
+  cachedRequestAsString: string
+): ReturnType<Request> {
+  const {
+    deps: { logger, dataStore },
+    idempotencyKey,
+    request,
+    operationName,
+    requestParams,
+    handleParamMismatch
+  } = args
+
+  let cachedRequest: CachedRequest
+
+  try {
+    cachedRequest = JSON.parse(cachedRequestAsString)
+  } catch (error) {
+    logger.error('Could not parse cache', { cachedRequestAsString })
+
+    await dataStore.delete(idempotencyKey)
+    return request()
+  }
+
+  if (
+    cachedRequest?.operationName !== operationName ||
+    !isEqual(
+      cachedRequest.requestParams,
+      JSON.parse(JSON.stringify(requestParams))
```

Comment:
`JSON.parse(JSON.stringify(requestParams))` required in order to make sure that the order of parameters passed in as graphql mutation input does not matter (in relation to the cached input), not sure if that is important or not